Fix password-protected product layout in blockified templates shows unnecessary product title and product excerpt

AhmarZaidi opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR checks for `core/post-title` and `core/post-excerpt` block and remove it's parent `core/column` block in the array reduce function while replacing the `single_product_template_blocks`.

The reason that we need to remove the parent `core/column` is that the `serialize_blocks()` gives error notices due to empty `core/column` block.


<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #50391 

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
| <img width="1462" alt="image" src="https://github.com/user-attachments/assets/bf5240b1-d95d-403a-a6ee-f9402a69091b" /> | <img width="1453" alt="image" src="https://github.com/user-attachments/assets/7a2d65c1-19c6-4d90-953e-997391da0a44" /> |


Comparing Classic Template and Blockified Template after fix:
| Classic Template (twenty-twnety-four) | Blockified Template (twenty-twnety-five) |
| ------ | ----- |
| <img width="1416" alt="image" src="https://github.com/user-attachments/assets/f96b7667-1f99-4987-ad3e-7d704d530375" /> | <img width="1453" alt="image" src="https://github.com/user-attachments/assets/7a2d65c1-19c6-4d90-953e-997391da0a44" /> |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

0. Make sure you are using the blockified Single Product template. You can go to Appearance > Site Editor > Templates > Single Product and, if you see the Classic Template Placeholder block, click on Transform into blocks.
1. Create or edit a product changing its visibility to Password protected.
2. Visit the product on the frontend and notice that the layout is similar to the classic template (product title and excerpt removed)

<img width="1453" alt="image" src="https://github.com/user-attachments/assets/7a2d65c1-19c6-4d90-953e-997391da0a44" />

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

AhmarZaidi committed:
fix: remove password protected product title and excerpt

AhmarZaidi committed:
add: changelog entry

AhmarZaidi committed:
Merge branch 'trunk' into fix/50391

Aljullu commented on plugins/woocommerce/src/Blocks/Templates/SingleProductTemplate.php:
```diff
@@ -142,9 +142,24 @@ private static function replace_first_single_product_template_block_with_passwor
 		// We want to replace the first single product template block with the password form. We also want to remove all other single product template blocks.
 		// This array doesn't contains all the blocks. For example, it missing the breadcrumbs blocks: it doesn't make sense replace the breadcrumbs with the password form.
 		$single_product_template_blocks = array( 'woocommerce/product-image-gallery', 'woocommerce/product-details', 'woocommerce/add-to-cart-form', 'woocommerce/product-meta', 'woocommerce/product-rating', 'woocommerce/product-price', 'woocommerce/related-products' );
```
Btw, no need to do in this PR, but while testing your PR I noticed there are some blocks that are not hidden in password-protected products. For example, _Related products_ (which is a _Product Collection_ block) is visible when it shouldn't:

![imatge](https://github.com/user-attachments/assets/e57f93a5-262e-4c10-8c35-5d0b1117407b)

In case you want to give it a try, I think we will need to add these extra blocks to the list:

* `woocommerce/add-to-cart-with-options`
* `woocommerce/product-gallery`
* `woocommerce/blockified-product-details`
* `woocommerce/product-collection`

As mentioned, not a blocker as it's not directly related to this task, but I noticed some blocks were missing from the list while reviewing your PR so I thought I would mention it anyway. :slightly_smiling_face:

Aljullu reviewed (commented):
Thanks for working on this, @AhmarZaidi!

> The reason that we need to remove the parent `core/column` is that the `serialize_blocks()` gives error notices due to empty `core/column` block.

Good finding. Apparently, that's an issue in the current implementation too. If the _Product Title_ and _Product Summary_ blocks are removed, we also end up with an empty column that causes some warnings to appear in the frontend.

I tested it and I could verify your implementation fixes that. But I worry the implemented solution might be a bit brittle. For example, if I add a _Spacer_ block before the _Product Title_ block, the warnings appear again:

![imatge](https://github.com/user-attachments/assets/90df0bc4-7f15-43a1-9473-46043e73fad1)

I wonder if a more reliable approach would be to remove the columns with no children right before this return:

https://github.com/woocommerce/woocommerce/blob/ce0ab84e71d2683140b063507ea8038ffd16c345/plugins/woocommerce/src/Blocks/Templates/SingleProductTemplate.php#L206-L211

Something along these lines, maybe:

```PHP
if ( count( $new_inner_blocks ) === 0 ) {
	return array(
		'blocks'              => $carry['blocks'],
		'html_block'          => null,
		'removed'             => true,
		'is_already_replaced' => $carry['is_already_replaced'],
	);
}

return array(
	'blocks'              => array_merge( $carry['blocks'], array( $block ) ),
	'html_block'          => null,
	'removed'             => false,
	'is_already_replaced' => $carry['is_already_replaced'],
);
```

AhmarZaidi committed:
Merge branch 'trunk' into fix/50391

AhmarZaidi committed:
add: check to remove empty inner blocks

AhmarZaidi commented:
Hey @Aljullu, Thanks for the review.

I've updated the logic according to the suggestions. Removing the empty inner blocks seems to be a better approach.

I've also added the missing woocommerce blocks to remove list.

> In case you want to give it a try, I think we will need to add these extra blocks to the list:
woocommerce/add-to-cart-with-options
woocommerce/product-gallery
woocommerce/blockified-product-details
woocommerce/product-collection

Aljullu reviewed (commented):
Thanks for the updates, @AhmarZaidi! I noticed the _Related Products_ block is still visible on the page:

![imatge](https://github.com/user-attachments/assets/9b79295e-0fd3-47dc-af1f-0c81579e0025)

I noticed in the tempalte it's in fact a pattern and not a _Product Collection_ block as I thought:

https://github.com/woocommerce/woocommerce/blob/a5cc5f3f90fcd4176daafffa9f3f32a00c77449f/plugins/woocommerce/templates/templates/blockified/single-product.html#L49

I wonder if we should add an extra check for this case. So when we check if `$block['blockName']` is in `$single_product_template_blocks` [here](https://github.com/woocommerce/woocommerce/blob/a5cc5f3f90fcd4176daafffa9f3f32a00c77449f/plugins/woocommerce/src/Blocks/Templates/SingleProductTemplate.php#L148), we could add an extra check if the block is a pattern and the `slug` is `woocommerce-blocks/related-products`. What do you think? I guess that way the _Related Products_ will be correctly hidden.

AhmarZaidi committed:
update: hide logic for related products pattern

Aljullu reviewed (commented):
Thanks for the updates, @AhmarZaidi! This looks perfect now:

Before | After
--- | ---
![imatge](https://github.com/user-attachments/assets/930dae44-e798-4862-8199-e918ac46d687) | ![imatge](https://github.com/user-attachments/assets/33899343-745e-462b-8c54-3666fa1c94c8)

<details>
<summary>Testing environment</summary>

* WordPress 6.8 local dev environment
* Plugins: Gutenberg, WooCommerce Beta Tester, WordPress Beta Tester
* Theme: Twenty Twenty-Five
* Store creation: April 15th 2025
* Store size: small
</details>

Apparently there is one test that needs to be updated before being able to merge:

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/php/src/Blocks/Templates/SingleProductTemplateTests.php#L308

If you need any help running the tests locally, let me know. :+1: The documentation can be found [here](https://github.com/woocommerce/woocommerce/blob/d905fc71e125b45d7bf2b7a700e0c6aaa9cb0c4f/plugins/woocommerce/tests/README.md#set-up-test-environment).

kalessil committed:
Merge branch 'trunk' into fix/50391

AhmarZaidi committed:
update: test for single product template

AhmarZaidi committed:
Merge branch 'fix/50391' of https://github.com/AhmarZaidi/woocommerce into fix/50391

AhmarZaidi committed:
Merge branch 'trunk' into fix/50391

Aljullu reviewed (approved):
Thanks for updating the test, @AhmarZaidi! LGTM :rocket: 

This improvement will ship in WooCommerce 9.9.
